### PR TITLE
[seeed_mr24hpc1] Fix frame parser length handling bugs

### PR DESCRIPTION
**Pull request by swoboda1337**
## What does this implement/fix?

Fix several bugs in the MR24HPC1 UART frame parser's data length handling:

1. **`FRAME_DATA_LEN_H`**: `sg_data_len_ = value * 256` was stored into a `uint8_t`, always truncating to 0 (dead code). Since the protocol max payload is 32 bytes, the high byte must be 0 for any valid frame. Changed to reject non-zero values directly.

2. **`FRAME_DATA_LEN_L`**: Zero-length data frames were accepted and entered `FRAME_DATA_BYTES`, where `sg_data_len_ -= 1` would underflow the `uint8_t` from 0 to 255, causing 256 bytes to be written into the 128-byte `sg_frame_buf_` (buffer overrun). Added `== 0` rejection. Also changed `+=` to direct assignment since the high byte contribution is always 0.

3. **`FRAME_DATA_BYTES`**: `sg_data_len_ <= 0` on `uint8_t` is misleading — it only matches exactly 0 since unsigned values are never negative. Changed to `--sg_data_len_ == 0` (pre-decrement with explicit zero check).

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Developer breaking change (an API change that could break external components)
- [ ] Undocumented C++ API change (removal or change of undocumented public methods that lambda users may depend on)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- N/A

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- N/A

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040/RP2350
- [ ] BK72xx
- [ ] RTL87xx
- [ ] LN882x
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# No config changes needed - existing configs work as before.
# This fixes frame parsing edge cases that could cause buffer overrun.
uart:
  rx_pin: GPIO16
  tx_pin: GPIO17
  baud_rate: 115200

seeed_mr24hpc1:
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/seeed_mr24hpc1/seeed_mr24hpc1.cpp [hunk 1]
```diff
@@ -297,19 +297,17 @@ void MR24HPC1Component::r24_split_data_frame_(uint8_t value) {
...
-        this->sg_data_len_ = 0;
         this->sg_recv_data_state_ = FRAME_IDLE;
         ESP_LOGD(TAG, "FRAME_DATA_LEN_H ERROR value:%x", value);
       }
       break;
     case FRAME_DATA_LEN_L:
-      this->sg_data_len_ += value;
-      if (this->sg_data_len_ > 32) {
+      this->sg_data_len_ = value;
+      if (this->sg_data_len_ == 0 || this->sg_data_len_ > 32) {
```
Comment (bdraco) [inline esphome/components/seeed_mr24hpc1/seeed_mr24hpc1.cpp:310]: nit: magic number

**Review by bdraco — APPROVED** [review]

**Comment by swoboda1337** [comment]
Thanks
